synapse/handlers/sync.py

```diff
@@ -1837,10 +1839,25 @@ async def _generate_sync_entry_for_rooms(
 
         since_token = sync_result_builder.since_token
 
+        # If all rooms are blocked, we can skip bits of processing.
+        block_all_rooms = (
+            sync_result_builder.sync_config.filter_collection.blocks_all_rooms()
+        )
+
+        # 0. If there are no rooms to return *and* we don't care about presence
+        # or device list updates, there's nothing to do.
+        if (
+            block_all_rooms
+            and block_all_presence_data
+            and not include_device_list_updates
+        ):
+            return set(), set(), set(), set()
+
```

Comment:
Maybe? The 2nd clause though would still come into effect for the device lists / presence stuff (i.e. the "if nothing has changed there's nothing to do"). Maybe by `joined` being empty though this would work fine without any other changes.

I'm hesitant to change that here!

Reply:
Very fair.

I hate the sync logic.

Reply:
OK, so I started combining the booleans as you asked in https://github.com/matrix-org/synapse/pull/14908#discussion_r1088216958; then realized it should just be hoisted to the caller.